spec/predicates/scai.md

```diff
@@ -0,0 +1,334 @@
+
+Type URI: scai/attribute-report/v0.1
+
+Version: 0.1
+
+Author: Marcela Melara <marcela.melara@intel.com>
+
+## Purpose
+
+The Software Supply Chain Attribute Integrity, or SCAI (pronounced "sky"), specification proposes a data
+format for capturing functional attribute and integrity information about software artifacts and their supply
+chain. SCAI data can be associated with executable binaries, statically- or dynamically-linked libraries,
+software packages, container images, software toolchains, and compute environments.
+
+Existing supply chain data formats do not capture any information about the security functionality or
+behavior of the resulting software artifact, nor do they provide sufficient evidence to support any claims of
+integrity of the supply chain processes they describe. The SCAI data format is designed to bridge this gap.
+
+For more details, see the [SCAI v0.1 specification document](https://arxiv.org/pdf/2210.05813.pdf)
+
+## Prerequisites
+
+In addition to the in-toto attestation framework,
+SCAI assumes that implementers have appropriate processes and tooling in place for
+capturing other types of software supply chain metadata, which can be extended to add support for SCAI.
+
+## Model
+
+SCAI metadata, referred to as an Attribute Assertion, describes functional attributes of a software
+artifact and its supply chain, capable of covering the full software stack of the
+toolchain that produced the artifact down to the hardware platform. SCAI Attribute Assertions include
+information about the conditions under which certain functional attributes arise, as well as (authenticated)
+evidence for the asserted attributes. The set of Assertions about a subject artifact and its producer
+is referred to as the **Attribute Report**.
+
+SCAI is intended to be implemented as part of an existing software supply chain attestation
+framework by software development tools or services (e.g., builders, CI/CD pipelines, software analysis tools)
+seeking to capture more granular information about the attributes and behavior of the software artifacts they
+produce. As such, we envision SCAI metadata being explicitly bound to, or embedded within, other metadata
+objects.
+
+## Schema
+
+The core metadata in SCAI is the Attribute Assertion. A collection of Attribute Assertions
+for a specific supply chain step or operation are issued together in a SCAI Attribute Report predicate.
+
+```
+{
+    "predicateType": "scai/attribute-report/v0.1",
+    "predicate": {
+        "subjectAttributes": [{
+            "attribute": "<ATTRIBUTE>",
+            "target": { // optional
+                "name": "<NAME>",
+                "digest": { "<ALGORITHM>": "<HEX VALUE>", ... },
+                "locationURI": "<URI>",
+                "objectType": "<OBJECT TYPE>" // optional
+            },
+            "conditions": { /* object */ }, // optional
+            "evidence": { /* object */ } // optional
+        }],
+        "producerAttributes": [{
```

Comment:
Should we not also capture the type/URI or ID of the producer to determine the schema of the evidence object? Based on our last call, we decided to tie the schema of the evidence object to the producer so that it does not remain arbitrary.

Reply:
I think we discussed this a little on today's call. The evidence isn't actually necessarily tied to the producer of the SCAI predicate, as it may come from a third-party auditor/scanner etc. 

That said, for the producer, I do think it would be useful to have some sort of `identifier` field for the producer. The main reason I haven't added it in explicitly, is because I think different use cases will identify the producer differently. For some, the producer's public key may be enough, others may want a URI, as you said. I wasn't sure how to encode these different use cases in a general manner. Maybe it could be dependent on the producer's type?

Reply:
@pxp928 I wanted to ping you about this again as I try to come up with a way to capture a type and ID for the producer. The producer of the subject is not necessarily the producer of the evidence, so we shouldn't encode this expectation into the predicate. But are there other use cases for having a type or ID for a producer? And what would the format of these fields be? 

I have a few different ways to think of a producer of a SCAI predicate:
* The producer is a build system that is making assertions about the attributes of subject artifacts. In this case, the type might be the URI of the build service (e.g. GHA), and the ID the specific build job.
* The producer is the actual compiler that built the subject. In case, the ID might describe the gcc binary itself. But I might also argue that we could describe the compiler in the `target` field of the `producer.attributes` object instead. In any case, what does the type/URI point to for a binary?

These are only two cases, but I would want to be able to capture these and other cases. And, I'm still trying to understand what a consumer of the predicate would do with this information.

Reply:
Here are a few examples. I also saw your example above with GUAC as the producer. Is GUAC a tool, a service or something else in this case?

Producer is GHA:
```
{
     "_type": "https://in-toto.io/Statement/v0.1",
     "subject": [{
        "name": "my-app",
        "digest": { "sha256": "78ab6a8..." }
    }],
    "predicateType": "scai/attribute-report/v0.1",
    "predicate": {
        "subjectAttributes": [{
             "attribute": "WITH_STACK_PROTECTION",
             "conditions": { "build-flags": "-fstack-protector" }
        }],
        "producer": {
            "type": "https://url-to-gha",
            "id": "https://example-builder.com/user/repo/actions/runs/runID"
            "attributes": [{ 
                "attribute": "ATTESTED_BUILD",
                "evidence": {
                    "name": "my-app-slsa-provenance",
                    "digest": { "sha256": "4567890..." },
                    "locationURI": "http://example.com/rekor-instance",
                    "objectType": "application/vnd.in-toto+json"
                }
            }]
        }    
    }
}
```

Producer is tool:
```
{
     "_type": "https://in-toto.io/Statement/v0.1",
     "subject": [{
        "name": "my-app",
        "digest": { "sha256": "78ab6a8..." }
    }],
    "predicateType": "scai/attribute-report/v0.1",
    "predicate": {
        "subjectAttributes": [{
             "attribute": "WITH_STACK_PROTECTION",
             "conditions": { "build-flags": "-fstack-protector" }
        }],
        "producer": {
            "type": "ELF binary",
            "id": {
                "name": "gcc9.3.0",
                "digest": { "sha256": "78ab6a8..." },
                "locationURI": "http://us.archive.ubuntu.com/ubuntu/pool/main/g/gcc-defaults/gcc_9.3.0 1ubuntu2_amd64.deb"
            }
            "attributes": [{ 
                "attribute": "ATTESTED_BUILD",
                "evidence": {
                    "name": "my-app-slsa-provenance",
                    "digest": { "sha256": "4567890..." },
                    "locationURI": "http://example.com/rekor-instance",
                    "objectType": "application/vnd.in-toto+json"
                }
            }]
        }    
    }
}
```

Reply:
So my comment was tied to when we spoke that the evidence field schema would be tied to a producer. In the latter meeting, we discussed this would not be the case. I agree the evidence filed would be different based on the type of producer (tool, service...etc.). Maybe we hold off for v0.1 but what you have for now should be sufficient:
```
        "producer": {
            "type": "<TYPE URI>",
```